### PR TITLE
AAP Integration Updates

### PR DESCRIPTION
**Pull request by ewsiegel**
Documentation and lab materials for adding a policy based governance hook into AAP automation upon deployment of sno2 via ZTP

Documentation Updates: 
- added integration-with-aap, configuring-aap, adding-automation
- added sections to existing pages lab-environment-introduction, check-deployment-is-finished, lab-environment

Lab Materials Updates: 
- added ansible-automation-platform-operator to mirrored operators
- added aap deployment to operators-deployment
- added AutomationHub and AutomationController CRD config to operators-config

Lab Environment Setup Updates: 
- added a new repo aap-integration-tools https://github.com/ewsiegel/aap-integration-tools to migrate to gitea server for managing Ansible playbooks via git SCM (see new line under Lab Environment > Configure Git Server)
- added a section at the bottom of lab-environment for uploading Ansible collections to a private Automation Hub during lab setup

Notes:
- aap integration part of the lab contains some ansible resources that depend on this repo: https://github.com/ewsiegel/aap-integration-tools
- changes to lab environment setup, still todo is update automation / deployment scripts for changes mentioned in Lab Environment Setup Updates above

### CODE REVIEW

**Comment by ewsiegel** [comment]
Update: https://github.com/ewsiegel/aap-integration-tools repo dependency now lives at https://github.com/RHsyseng/5g-ran-lab-aap-integration-tools and corresponding lab setup and documentation updates have been made in commit 07195d806e5078e849b2502d3f688229a9db3eec

File: documentation/modules/ROOT/pages/integration-with-aap.adoc [hunk 1]
```diff
@@ -0,0 +1,176 @@
+= Integration with Ansible Automation Platform (AAP)
+include::_attributes.adoc[]
+:profile: 5g-ran-lab
+
+
+[#aap-overview]
+== AAP Overview
+
+link:https://www.redhat.com/en/technologies/management/ansible[Red Hat Ansible Automation Platform] allows for the configuration and management of systems at scale through simple IT automation. If unfamiliar with AAP, see the link:https://docs.ansible.com/ansible/latest/index.html[AAP Docs]. In this section, we will discuss the need for AAP integration with 5G RAN deployments on OpenShift.
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/integration-with-aap.adoc]: Change the link from community docs to Red Hat docs. i.e: https://access.redhat.com/documentation/en-us/red_hat_ansible_automation_platform/2.4

File: documentation/modules/ROOT/pages/integration-with-aap.adoc [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+[#augmenting-openshift]
+=== Augmenting Openshift with Ansible Automation
+
+As discussed in previous sections, we leverage RHACM and Openshift to configure our clusters through custom resource definitions and policy based governance. These definitions define the desired state of the clusters, and upon deployment and beyond, Openshift works to bring the clusters to compliance with our policies. This is known as a _declarative_ model of configuration, and alongside GitOps and our ZTP workflow, allows for simplicity and flexibility when managing a large number clusters at scale.
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/integration-with-aap.adoc]: OpenShift

File: documentation/modules/ROOT/pages/integration-with-aap.adoc [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+For a deeper dive, see link:https://www.ansible.com/blog/openshift-and-ansible-bridging-the-automation-gap-for-5g-and-beyond-networks-part-1[OpenShift and Ansible: Bridging the Automation Gap for 5G and Beyond Networks - Part 1] on the Ansible blog.
+
+[#intro-to-aap-operator]
+=== Introduction to the AAP Operator
+
+The AAP operator provides us with the necessary components to host both Ansible collections (through the Automation Hub) to leverage Ansible's diverse distribution platform and Ansible automation objects (through the Automation Controller) natively through Openshift. The operator offers the custom resource definitions AutomationHub and AutomationController, whose functions are defined below.
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/integration-with-aap.adoc]: OpenShift

File: documentation/modules/ROOT/pages/integration-with-aap.adoc [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+[#aap-components]
+== Defining AAP Components
+
+[#automation-hub]
+=== Automation Hub
+
+The OpenShift AAP operator's Automation Hub, defined by the custom resource definition AutomationHub, is a downstream version of link:https://github.com/ansible/galaxy_ng[Galaxy NG] which allows for the hosting of Ansible collections on a private distribution server. Collections are a distribution format for Ansible objects including playbooks, roles, modules, and plugins. We store all necessary Ansible collections through an Automation Hub deployed on OpenShift, allowing all dependencies to be hosted and managed through our local cluster. We can upload collections, manage collections, and interact with the Automation Hub through API or through its web UI deployed at automation-hub-aap.apps.hub.5g-deployment.lab.
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/integration-with-aap.adoc]: Convert automation-hub-aap.apps.hub.5g-deployment.lab to link:

https://automation-hub-aap.apps.hub.5g-deployment.lab(https://automation-hub-aap.apps.hub.5g-deployment.lab)

File: documentation/modules/ROOT/pages/integration-with-aap.adoc [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+An example AutomationHub CRD is shown below.
+
+[source,bash]
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/integration-with-aap.adoc]: You can use source,yaml to get proper coloring.

File: documentation/modules/ROOT/pages/integration-with-aap.adoc [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+=== Automation Controller
+
+The Openshift AAP operator's Automation Controller, defined by the custom resource definition AutomationController, deploys an Ansible Automation Controller (formerly Ansible Tower) in our OpenShift cluster. Downstream of the link:https://github.com/ansible/awx[AWX Project] in link:https://github.com/ansible/awx-operator[operator form], our deployed Automation Controller provides a user interface, REST API, and task engine for executing Ansible jobs. The Automation Controller's functionality includes managing Ansible automation objects including projects, inventories, playbooks, roles, credentials, and job templates, as well as the responsibility of invoking and monitoring the execution of Ansible jobs. 
+
+Details of the contents of an Ansible job will be covered as we craft our automation template. At a high level, an Ansible _job_ is a running instance of a _job template_, which contains a logical collection of Ansible _playbooks_ in the form of a _project_, a list of managed hosts in the form of an _inventory_, and any extra variables we want to pass to the job. 
+
+An example AutomationController CRD is shown below.
+
+[source,bash]
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/integration-with-aap.adoc]: source,yaml

File: documentation/modules/ROOT/pages/integration-with-aap.adoc [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+== Workflow Overview
+
+Now that we are familiar with AAP in OpenShift and the necessary components for integration, let's walk through an Ansible integration workflow.
+
+[#general-workflow]
+=== General AAP Integration Workflow
+
+At a high level, the steps to AAP integration with 5G RAN ZTP deployments are as follows:
+
+* Mirror AAP operator to disconnected registry (outside the scope of this lab, see the _Ansible Automation Platform mirroring_ section of this link:https://www.ansible.com/blog/openshift-and-ansible-bridging-the-automation-gap-for-5g-and-beyond-networks-part-1[blog])
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/integration-with-aap.adoc]: If there are official docs for this we want to point to them instead.

File: documentation/modules/ROOT/pages/integration-with-aap.adoc [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+* Install AAP operator, deploy AutomationHub and AutomationController (already running on OpenShift cluster, see the _Ansible Automation Platform operator installation_ section of this link:https://www.ansible.com/blog/openshift-and-ansible-bridging-the-automation-gap-for-5g-and-beyond-networks-part-1[blog] and the CRD examples for link:integration-with-aap.html#automation-hub[AutomationHub] and link:integration-with-aap.html#automation-controller[AutomationController])
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/integration-with-aap.adoc]: If there are official docs for this we want to point to them instead.

File: documentation/modules/ROOT/pages/integration-with-aap.adoc [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+* Upload necessary Ansible collections to Automation Hub (already uploaded, see uploaded collections at automation-hub-aap.apps.hub.5g-deployment.lab and instructions for this process in the link:lab-environment.html#upload-collections-to-aap-hub[Lab Environment] section)
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/integration-with-aap.adoc]: We will do this in our automation. Pointing users to Lab Environment docs is not advised, it can be confusing (they're not supposed to run these when they get an env from demo.redhat.com)

File: documentation/modules/ROOT/pages/integration-with-aap.adoc [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+* Configure Automation Controller with Ansible automation resources
+* Install MultiClusterHub add-ons
+* Deploy spoke clusters
+* Deploy ManagedServiceAccount, Policy, PlacementBinding, PlacementRule, and PolicyAutomation for a governance-based hook into AAP
+* Spoke cluster(s) reach state where Policy applies, Ansible job is triggered by PolicyAutomation on each cluster, Ansible job completes successfully, Policy moves to compliant for that cluster
+
+This lab will cover the process starting with configuring the Automation Controller and ending with a successful Ansible job and a compliant automation Policy for a deployed cluster.
+
+[#lab-workflow]
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/integration-with-aap.adoc:68]: As I put in the other docs, I think this is outside the scope of the lab and should be done in the automation. No need to explain the specific steps to the user since they will not be doing them.

File: documentation/modules/ROOT/pages/lab-environment-introduction.adoc [hunk 11]
```diff
@@ -38,3 +38,39 @@ Credentials for the OCP cluster (htpassidp provider):
...
 
+[#automation-hub-and-controller]
+== AAP Automation Hub and Controller
+
+The Automation Hub can be reached at https://automation-hub-aap.apps.hub.5g-deployment.lab
+
+Credentials for Automation Hub:
+
+* Username: `admin`
+* Password: Retrieve the password by logging into oc and running:
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/lab-environment-introduction.adoc]: We will send this password in the email. Do we know if AAP can be integrated with the cluster oauth?

File: documentation/modules/ROOT/pages/lab-environment-introduction.adoc [hunk 12]
```diff
@@ -38,3 +38,39 @@ Credentials for the OCP cluster (htpassidp provider):
...
+
+[.console-input]
+[source,bash]
+-----
+oc -n aap get secret automation-hub-admin-password -o jsonpath='{.data.password}' | base64 -d
+-----
+
+The Automation Controller can be reached at https://automation-aap.apps.hub.5g-deployment.lab
+ 
+Credentials for Automation Controller:
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/lab-environment-introduction.adoc:53]: We will send this password in the email. Do we know if AAP can be integrated with the cluster oauth?

File: documentation/modules/ROOT/pages/lab-environment-introduction.adoc [hunk 13]
```diff
@@ -38,3 +38,39 @@ Credentials for the OCP cluster (htpassidp provider):
...
+* Username: `admin`
+* Password: Retrieve the password by logging into oc and running: 
+
+[.console-input]
+[source,bash]
+-----
+oc -n aap get secret automation-admin-password -o jsonpath='{.data.password}' | base64 -d
+-----
+
+NOTE: If the above URLs are not reachable, add the following entry to your `/etc/hosts`:
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/lab-environment-introduction.adoc]: We don't want to put this here.

File: documentation/modules/ROOT/pages/configuring-aap.adoc [hunk 14]
```diff
@@ -0,0 +1,185 @@
+= Configuring Ansible Automation Platform
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/configuring-aap.adoc]: I feel like all these sections should be removed and instead get all these steps in our automation. The lab is about 5G RAN OCP Deployments, configuring AAP IMO is outside the scope. I like the idea of leveraging AAP in the workflow. I'd say let's figure out how we can automate all this stuff so it's ready when the user access the lab.

File: documentation/modules/ROOT/pages/check-deployment-is-finished.adoc [hunk 15]
```diff
@@ -65,5 +65,26 @@ ztp-policies.common-subscription-policies   inform               Compliant
...
+
+Now that SNO deployment is finished, automation should run as soon as the cluster is marked with the `ztp-done` label. We can easily check the labels on our `sno2` cluster via the OpenShift web console, under the `Infrastructure` > `Clusters` tab.
+
+image::sno2-labels.png[SNO2 Labels]
+
+If the `ztp-done` label is present, we can check to see whether our AAP automation ran successfully by monitoring the `sno2` cluster's policy governance. If the `monitor-ztp-done` policy is applied to `sno2` and marked as compliant, then we know that our automation has run successfully.
+
+NOTE: Deployment of the hub cluster was not part of this lab, but `local-cluster` was also deployed via a ZTP workflow, so it will have a `ztp-done` label and our policy will run automation on `local-cluster` as well. This model of maintaing one automation policy for day 2 operations for N clusters makes utilizng ZTP and AAP automation together in a 5G RAN environment great at scale while also maintaining configuration control, as controlling which clusters run automation is as simple as adjusting the automation policy's PlacementRule.
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/check-deployment-is-finished.adoc]: typo: utilizng

File: documentation/modules/ROOT/pages/check-deployment-is-finished.adoc [hunk 16]
```diff
@@ -65,5 +65,26 @@ ztp-policies.common-subscription-policies   inform               Compliant
...
+
+image::automation-policy.png[Monitor-ZTP-Done Policy]
+
+Lastly, to view status and logs for the actual Ansible Controller job that ran, we can navigate to automation-aap.apps.hub.5g-deployment.lab, click on the "Jobs" tab, and view the latest job that ran.
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/check-deployment-is-finished.adoc]: Make automation-aap.apps.hub.5g-deployment.lab a link.

File: documentation/modules/ROOT/pages/adding-automation.adoc [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+
+[#enabling-managedserviceaccount]
+== Enabling the ManagedServiceAccount Add-On
+
+We have already installed ManagedServiceAccount to our MultiClusterHub, but we must still enable and deploy a unique ManagedServiceAccount on each spoke cluster that we wish to run automation on. We accomplish this through the follow OpenShift CRDs. Notice these resources exist in the `sno2` namespace.
+
+[.console-input]
+[source,bash]
+-----
+cat << EOF | oc apply -f -
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/adding-automation.adoc]: oc command is missing the target kubeconfig

File: documentation/modules/ROOT/pages/adding-automation.adoc [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+apiVersion: addon.open-cluster-management.io/v1alpha1
+kind: ManagedClusterAddOn
+metadata:
+  name: managed-serviceaccount
+  namespace: sno2
+spec:
+  installNamespace: open-cluster-management-agent-addon
+EOF
+
+cat << EOF | oc apply -f -
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/adding-automation.adoc]: oc command is missing the target kubeconfig

File: documentation/modules/ROOT/pages/adding-automation.adoc [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+
+[#creating-governance-hook]
+== Crafting a Governance Hook into AAP
+
+In order to create our governance hook, we will need to first create a Policy that will monitor for the `ztp-done` label and trigger an Ansible job in the AAP Controller. This Policy should stay in a violating status until the automation job completes successfully and the `ztp-day2-automation` namespace is created on our `sno2` cluster. In order for this Policy to bind to clusters with the `ztp-done` label, we must also create the relevant PlacementRule and a PlacementBinding to bind the Policy to this PlacementRule. Notice these resources exist in the `ran-lab-automation` namespace.
+
+[.console-input]
+[source,bash]
+-----
+cat << EOF | oc -n ran-lab-automation apply -f -
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/adding-automation.adoc:92]: oc command is missing the target kubeconfig

File: documentation/modules/ROOT/pages/adding-automation.adoc [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+      - {key: ztp-done, operator: Exists}
+EOF
+-----
+
+Lastly, in order to actually trigger the Ansible job, we create an AnsibleJob based PolicyAutomation resource also in the `ran-lab-automation` namespace alongside our defined Policy. When the Ansible job is triggered, RHACM automatically injects an extra variable `target_clusters`, which is a list of clusters to run automation on based on our policy's PlacementRule.
+
+[.console-input]
+[source,bash]
+-----
+cat << EOF | oc -n ran-lab-automation apply -f -
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/adding-automation.adoc]: oc command is missing the target kubeconfig

File: documentation/modules/ROOT/pages/integration-with-aap.adoc (hunk 5 above)
Comment (mvazquezc) [reply documentation/modules/ROOT/pages/integration-with-aap.adoc]: After having read the other sections, if we pre-configure the whole AAP instance I don't think this will be relevant for this lab anymore.

File: documentation/modules/ROOT/pages/integration-with-aap.adoc (hunk 6 above)
Comment (mvazquezc) [reply documentation/modules/ROOT/pages/integration-with-aap.adoc]: After having read the other sections, if we pre-configure the whole AAP instance I don't think this will be relevant for this lab anymore.

File: documentation/modules/ROOT/pages/integration-with-aap.adoc [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+=== This Lab's AAP Integration Workflow
+
+This section will overview our example AAP integration workflow in this lab. If unfamiliar with the general AAP integration worflow, see the link:integration-with-aap.html#general-workflow[above section].
+
+In this lab, we will deploy a SNO cluster through the ZTP pipeline and await the completion of the SNO deployment, which will mark the SNO cluster with the _ztp-done_ label. The _ztp-done_ label will trigger an Ansible job that creates the namespace _ztp-day2-automation_ within our deployed SNO cluster. Assuming AAP is installed and deployed, and the necessary collections exist in our deployed Automation Hub, the steps to configure this integration is as follows:
+
+* *Configure Automation Controller with Ansible automation resources*
```
Comment (mvazquezc) [inline documentation/modules/ROOT/pages/integration-with-aap.adoc]: As I put in the other docs, I think this is outside the scope of the lab and should be done in the automation. No need to explain the specific steps to the user since they will not be doing them.

**Review by mvazquezc — CHANGES_REQUESTED** [review]
Hey Ethan,

I added a few comments.

File: documentation/modules/ROOT/pages/integration-with-aap.adoc (hunk 5 above)
Comment (ewsiegel) [reply documentation/modules/ROOT/pages/integration-with-aap.adoc]: Will look into pre-configuring AAP to remove the need for setup related documentation and focus on functionality.

File: documentation/modules/ROOT/pages/configuring-aap.adoc (hunk 14 above)
Comment (ewsiegel) [reply documentation/modules/ROOT/pages/configuring-aap.adoc]: We can definitely automate the steps in the "#configuring-automation-controller" and "#installing-mch-addons" sections. This would include general Automation Controller setup (AutomationController instance with existing subscription loaded, generate authentication tokens), the actual creation of the Ansible project and job template, and patching MultiClusterHub with certain add-ons (which can actually just be included in original MultiClusterHub config in my understanding).

We can then move the OpenShift CRDs applied in the "#adding-supporting-objects" section to "adding-automation.adoc" instead, since this section involves simple OpenShift resources that are critical for the Policy creation.

Would this change make sense?

Comment (mvazquezc) [reply documentation/modules/ROOT/pages/configuring-aap.adoc]: Sounds good to me. The Tower token should be generated in our automation and sent via email. I want to avoid users having to access the hypervisor to run cat or something.

File: documentation/modules/ROOT/pages/lab-environment-introduction.adoc (hunk 11 above)
Comment (ewsiegel) [reply documentation/modules/ROOT/pages/lab-environment-introduction.adoc]: TODO: demo environment automation and include Automation Hub password, Automation hub token, Automation Controller password, Automation Controller token in email
